bakery/codec.go

```diff
@@ -23,109 +28,158 @@ type caveatId struct {
+func encodeJSONCaveatId(key *KeyPair, ci caveatInfo) ([]byte, error) {
 	var nonce [NonceLen]byte
 	if _, err := rand.Read(nonce[:]); err != nil {
-		return nil, fmt.Errorf("cannot generate random number for nonce: %v", err)
+		return nil, errgo.Notef(err, "cannot generate random number for nonce")
 	}
 	plain := caveatIdRecord{
-		RootKey:   rootKey,
-		Condition: condition,
+		RootKey:   ci.RootKey,
+		Condition: ci.Predicate,
 	}
 	plainData, err := json.Marshal(&plain)
 	if err != nil {
-		return nil, fmt.Errorf("cannot marshal %#v: %v", &plain, err)
+		return nil, errgo.Notef(err, "cannot marshal %#v", &plain)
 	}
-	sealed := box.Seal(nil, plainData, &nonce, thirdPartyPub.boxKey(), enc.key.Private.boxKey())
-	return &caveatId{
-		ThirdPartyPublicKey: thirdPartyPub,
-		FirstPartyPublicKey: &enc.key.Public,
+	sealed := key.seal(ci.PeerPublicKey, &nonce, plainData)
+	id := caveatId{
+		ThirdPartyPublicKey: ci.PeerPublicKey,
+		FirstPartyPublicKey: &key.Public,
 		Nonce:               nonce[:],
 		Id:                  base64.StdEncoding.EncodeToString(sealed),
-	}, nil
+	}
+	data, err := json.Marshal(id)
+	if err != nil {
+		return nil, errgo.Notef(err, "cannot marshal %#v", id)
+	}
+	buf := make([]byte, base64.StdEncoding.EncodedLen(len(data)))
+	base64.StdEncoding.Encode(buf, data)
+	return buf, nil
 }
 
-// boxDecoder decodes caveat ids for third-party service that were encoded to
-// the third-party with authenticated public key encryption compatible with
-// NaCl box.
-type boxDecoder struct {
-	key *KeyPair
+const (
+	v0ThirdPartyPublicKeyPrefixOffset = 1
+	v0FirstPartyPublicKeyOffset       = v0ThirdPartyPublicKeyPrefixOffset + 4
+	v0NonceOffset                     = v0FirstPartyPublicKeyOffset + KeyLen
+	v0EncryptedPartOffset             = v0NonceOffset + NonceLen
+
+	v0RootKeyOffset    = 1
+	v0CaveatTextOffset = v0RootKeyOffset + 24
+)
+
+// encodeCaveatIdV0 creates a version 0 third-party caveat.
+func encodeCaveatIdV0(key *KeyPair, ci caveatInfo) ([]byte, error) {
+	var nonce [NonceLen]byte
+	if _, err := rand.Read(nonce[:]); err != nil {
+		return nil, errgo.Notef(err, "cannot generate random number for nonce")
+	}
+	sealed := encodeSecretPartV0(key, ci, &nonce)
+	data := make([]byte, v0EncryptedPartOffset+len(sealed))
+	data[0] = 0
+	copy(data[v0ThirdPartyPublicKeyPrefixOffset:v0FirstPartyPublicKeyOffset], ci.PeerPublicKey.Key[:])
+	copy(data[v0FirstPartyPublicKeyOffset:v0NonceOffset], key.Public.Key[:])
+	copy(data[v0NonceOffset:v0EncryptedPartOffset], nonce[:])
+	copy(data[v0EncryptedPartOffset:], sealed)
+	return data, nil
 }
 
-// newBoxDecoder creates a new BoxDecoder using the given key pair.
-func newBoxDecoder(key *KeyPair) *boxDecoder {
-	return &boxDecoder{
-		key: key,
+// encodeSEcretPartV0 creates an encrypted secret part of the third party
+// caveat at version 0.
+func encodeSecretPartV0(key *KeyPair, ci caveatInfo, nonce *[NonceLen]byte) []byte {
+	data := make([]byte, len(ci.RootKey)+len(ci.Predicate)+1)
```

Comment:
Less obvious a win here, but probably nice to use the same append idiom I suggested above:

    data := make([]byte, 0, 1 + len(ci.RootKey) + len(ci.Predicate))
    data = append(data, 0)    // version
    data = append(data, ci.RootKey...)
    data = key.seal(data, ci.Predicate...)
    return key.seal(ci.PeerPublicKey, nonce, data)